File: webui/model/policy/policy.php
<?php


class ModelPolicyPolicy extends Model {

   public function get_policies($s = '') {
      if($s) {
         $query = $this->db->query("SELECT * FROM " . TABLE_POLICY . " WHERE `from` LIKE ? OR `to` LIKE ? OR subject LIKE ? ORDER BY domain, id", array('%' . $s . '%', '%' . $s . '%', '%' . $s . '%'));
      } else {
         $query = $this->db->query("SELECT * FROM " . TABLE_POLICY . " ORDER BY id");
      }

      if(isset($query->rows)) { return $query->rows; }

      return array();
   }


   public function get_policy($id = 0) {
      $query = $this->db->query("SELECT * FROM " . TABLE_POLICY . " WHERE id=?", array($id));

      if(isset($query->row)) { return $query->row; }

      return array();
   }


   public function add($data = array()) {
      $query = $this->db->query("INSERT INTO " . TABLE_POLICY . " (`name`,`deliver_infected_email`,`silently_discard_infected_email`,`use_antispam`,`spam_subject_prefix`,`max_message_size_to_filter`,`surbl_domain`,`spam_overall_limit`,`spaminess_oblivion_limit`,`replace_junk_characters`, `penalize_images`, `penalize_embed_images`, `penalize_octet_stream`, `training_mode`, `store_emails`, `store_only_spam`, `message_from_a_zombie`, `smtp_addr`, `smtp_port`) VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)", array($data['name'], $data['deliver_infected_email'], $data['silently_discard_infected_email'], $data['use_antispam'], $data['spam_subject_prefix'], $data['max_message_size_to_filter'], $data['surbl_domain'], $data['spam_overall_limit'], $data['spaminess_oblivion_limit'], $data['replace_junk_characters'], $data['penalize_images'], $data['penalize_embed_images'], $data['penalize_octet_stream'], $data['training_mode'], $data['store_emails'], $data['store_only_spam'], $data['message_from_a_zombie'], $data['smtp_addr'], $data['smtp_port']));

      return $this->db->countAffected();
   }


   public function update($data = array()) {

      if(count($data) < 5 || $data['id'] <= 0 || !is_numeric($data['id'])){
         return 0;
      }

      $query = $this->db->query(
         "UPDATE " . TABLE_POLICY . " SET `name`=?,`deliver_infected_email`=?,`silently_discard_infected_email`=?,`use_antispam`=?,`spam_subject_prefix`=?,`max_message_size_to_filter`=?,`surbl_domain`=?,`spam_overall_limit`=?,`spaminess_oblivion_limit`=?,`replace_junk_characters`=?, `penalize_images`=?, `penalize_embed_images`=?, `penalize_octet_stream`=?, `training_mode`=?, `store_emails`=?, `store_only_spam`=?, `message_from_a_zombie`=?, `smtp_addr`=?, `smtp_port`=? WHERE id=?", array($data['name'], $data['deliver_infected_email'], $data['silently_discard_infected_email'], $data['use_antispam'], $data['spam_subject_prefix'], $data['max_message_size_to_filter'], $data['surbl_domain'], $data['spam_overall_limit'], $data['spaminess_oblivion_limit'], $data['replace_junk_characters'], $data['penalize_images'], $data['penalize_embed_images'], $data['penalize_octet_stream'], $data['training_mode'], $data['store_emails'], $data['store_only_spam'], $data['message_from_a_zombie'], $data['smtp_addr'], $data['smtp_port'], $data['id']));

      $rc = $this->db->countAffected();

      LOGGER("update policy: " . $data['name'] . " (id=" . (int)$data['id'] . ") (rc=$rc)");

      return $rc;
   }

   public function remove($id = 0) {
      $query = $this->db->query("DELETE FROM " .  TABLE_POLICY . " WHERE id=?", array($id));
      return $this->db->countAffected();
   }


   public function fix_post_data() {

      foreach(array(
                     'deliver_infected_email',
                     'silently_discard_infected_email',
                     'use_antispam',
                     'replace_junk_characters',
                     'penalize_images',
                     'penalize_embed_images',
                     'penalize_octet_stream',
                     'store_emails',
                     'store_only_spam'
               )
               as $a) {

         if(!isset($this->request->post[$a])) { $this->request->post[$a] = 0; }
         else { $this->request->post[$a] = 1; }
      }

   }


}

?>

File: src/policy.c
/*
 * policy.c, SJ
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <clapf.h>


int get_policy(struct session_data *sdata, struct __config *cfg, struct __config *my_cfg){
   int rc=0;
   float f;
   char f1[MAXVAL], f2[MAXVAL];
   struct sql sql;

   /*
    * in case of smtp mode don't query unless we have a single recipient
    */

   if(cfg->server_mode == SMTP_MODE && sdata->num_of_rcpt_to != 1) return rc;

   if(prepare_sql_statement(sdata, &sql, SQL_PREPARED_STMT_QUERY_POLICY) == ERR) return rc;

   p_bind_init(&sql);

   sql.sql[sql.pos] = (char *)&(sdata->policy_group); sql.type[sql.pos] = TYPE_LONG; sql.pos++;

   if(p_exec_stmt(sdata, &sql) == ERR) goto ENDE;

   p_bind_init(&sql);

   sql.sql[sql.pos] = (char *)&(my_cfg->deliver_infected_email); sql.type[sql.pos] = TYPE_LONG; sql.len[sql.pos] = sizeof(int); sql.pos++;
   sql.sql[sql.pos] = (char *)&(my_cfg->silently_discard_infected_email); sql.type[sql.pos] = TYPE_LONG; sql.len[sql.pos] = sizeof(int); sql.pos++;
   sql.sql[sql.pos] = (char *)&(my_cfg->use_antispam); sql.type[sql.pos] = TYPE_LONG; sql.len[sql.pos] = sizeof(int); sql.pos++;
   sql.sql[sql.pos] = my_cfg->spam_subject_prefix; sql.type[sql.pos] = TYPE_STRING; sql.len[sql.pos] = MAXVAL-1; sql.pos++;
   sql.sql[sql.pos] = (char *)&(my_cfg->max_message_size_to_filter); sql.type[sql.pos] = TYPE_LONG; sql.len[sql.pos] = sizeof(int); sql.pos++;
   sql.sql[sql.pos] = my_cfg->surbl_domain; sql.type[sql.pos] = TYPE_STRING; sql.len[sql.pos] = MAXVAL-1; sql.pos++;
   sql.sql[sql.pos] = &f1[0]; sql.type[sql.pos] = TYPE_STRING; sql.len[sql.pos] = MAXVAL-1; sql.pos++;
   sql.sql[sql.pos] = &f2[0]; sql.type[sql.pos] = TYPE_STRING; sql.len[sql.pos] = MAXVAL-1; sql.pos++;
   sql.sql[sql.pos] = (char *)&(my_cfg->replace_junk_characters); sql.type[sql.pos] = TYPE_LONG; sql.len[sql.pos] = sizeof(int); sql.pos++;
   sql.sql[sql.pos] = (char *)&(my_cfg->penalize_images); sql.type[sql.pos] = TYPE_LONG; sql.len[sql.pos] = sizeof(int); sql.pos++;
   sql.sql[sql.pos] = (char *)&(my_cfg->penalize_embed_images); sql.type[sql.pos] = TYPE_LONG; sql.len[sql.pos] = sizeof(int); sql.pos++;
   sql.sql[sql.pos] = (char *)&(my_cfg->penalize_octet_stream); sql.type[sql.pos] = TYPE_LONG; sql.len[sql.pos] = sizeof(int); sql.pos++;
   sql.sql[sql.pos] = (char *)&(my_cfg->training_mode); sql.type[sql.pos] = TYPE_LONG; sql.len[sql.pos] = sizeof(int); sql.pos++;
   sql.sql[sql.pos] = (char *)&(my_cfg->store_emails); sql.type[sql.pos] = TYPE_LONG; sql.len[sql.pos] = sizeof(int); sql.pos++;
   sql.sql[sql.pos] = (char *)&(my_cfg->store_only_spam); sql.type[sql.pos] = TYPE_LONG; sql.len[sql.pos] = sizeof(int); sql.pos++;
   sql.sql[sql.pos] = (char *)&(my_cfg->message_from_a_zombie); sql.type[sql.pos] = TYPE_LONG; sql.len[sql.pos] = sizeof(int); sql.pos++;
   sql.sql[sql.pos] = my_cfg->smtp_addr; sql.type[sql.pos] = TYPE_STRING; sql.len[sql.pos] = MAXVAL-1; sql.pos++;


   p_store_results(&sql);

   if(p_fetch_results(&sql) == OK){
      f = atof(f1); if(f > 0.1) my_cfg->spam_overall_limit = f;
      f = atof(f2); if(f > 0.1) my_cfg->spaminess_oblivion_limit = f;

      if(cfg->verbosity >= _LOG_DEBUG) syslog(LOG_PRIORITY, "%s: policy settings: %d/%d/%d/%s/%d/%s/%.4f/%.4f/%d/%d/%d/%d/%d/%d/%d/%d/%s",
                                              sdata->ttmpfile, my_cfg->deliver_infected_email, my_cfg->silently_discard_infected_email, my_cfg->use_antispam,
                                              my_cfg->spam_subject_prefix, my_cfg->max_message_size_to_filter, my_cfg->surbl_domain, my_cfg->spam_overall_limit,
                                              my_cfg->spaminess_oblivion_limit, my_cfg->replace_junk_characters, my_cfg->penalize_images, my_cfg->penalize_embed_images,
                                              my_cfg->penalize_octet_stream, my_cfg->training_mode, my_cfg->store_emails, my_cfg->store_only_spam, my_cfg->message_from_a_zombie,
                                              my_cfg->smtp_addr);
   }

   p_free_results(&sql);


ENDE:
   close_prepared_statement(&sql);

   return rc;
}
